cupy/random/_generator.py

```diff
@@ -1150,26 +1170,46 @@ def randint(self, low, high=None, size=None, dtype=int):
             - :func:`cupy.random.randint` for full documentation
             - :meth:`numpy.random.RandomState.randint`
         """
-        if high is None:
-            lo = 0
-            hi1 = int(low) - 1
+        if not numpy.isscalar(low):
+            low = cupy.asarray(low)
+            if high is None:
+                lo = cupy.zeros_like(low)
+                hi = low - 1
+            else:
+                lo = low
+                hi = cupy.asarray(high) - 1
+
+            if size is None:
+                size = cupy.broadcast(lo, hi).shape
+
+            diff = hi - lo
+            total_elems = functools.reduce(operator.mul, size, 1)
+            out = self._interval(diff.flatten(), total_elems)
+            out = cupy.reshape(out, size)
+            return out + lo
```

Comment:
I get a float64 result from `cp.random.randint([0],[10])`, which comes from `out`(uint64) + `lo` (int64). We'd like to have the same dtype as numpy here. Would you take care of this?
```py
>>> cp.random.randint([0],[10])
array([5.])  # dtype=cp.float64
>>> np.random.randint([0],[10])
array([8])  # dtype=np.int64
```

It seems that numpy does the work in `numpy.random._bounded_integers._rand_int64_broadcast`. I think it's defined in the following file as `_rand_{{nptype}}_broadcast`, which is written in Cython with maybe numpy specific source code templating. Cython is a kind of compiler that translates Python-like syntax to C, which CuPy also uses.
https://github.com/numpy/numpy/blob/main/numpy/random/_bounded_integers.pyx.in

Reply:
Just for clarification purposes, would you like to mask the inputs as in: https://github.com/numpy/numpy/blob/3b8dca3d0588048c425488a5e39ceddf815d9138/numpy/random/_bounded_integers.pyx.in#L139? Wouldn't a `astype` cast fix the issues?

Reply:
At least, `float32`/`float64` can not represent all values that `int32`/`int64` do. For example, `float64` has only 52 bits for fraction.

Reply:
Perhaps, we can use `int64` instead of `uint64` in `_interval` if the supplied dtype is so?